File: list.c
#include <stdio.h>
#include <stdlib.h>

typedef struct item {
	int value;
	struct item *prev, *next;
} item;

typedef struct list {
	size_t length;
	item *head, *tail;
} list;

item *new_item(int value) {
	item *it = (item *)malloc(sizeof(item));
	it->value = value;
	it->prev = NULL;
	it->next = NULL;
	return it;
}

list *new_list(size_t length) {
	if (length == 0) {
		return NULL;
	}

	list *ls = (list *)malloc(sizeof(list));
	ls->length = length;

	item *it = new_item(0);
	ls->head = it;
	for (int i = 0; i < length-1; i++) {
		it->next = new_item(0);
		it->next->prev = it;
		it = it->next;
	}
	ls->tail = it;

	return ls;
}

void expand(list *ls, size_t length) {
	ls->length += length;

	item *it = ls->tail;
	for (int i = 0; i < length; i++) {
		it->next = new_item(0);
		it->next->prev = it;
		it = it->next;
	}
	ls->tail = it;
}

item *set(list *ls, int index, int value) {
	if (index > ls->length-1) {
		return NULL;
	}

	item *it = ls->head;
	for (int i = 0; i < index; i++) {
		it = it->next;
	}
	it->value = value;
}

int get(list *ls, int index) {
	if (index > ls->length-1) {
		return NULL;
	}

	item *it = ls->head;
	for (int i = 0; i < index; i++) {
		it = it->next;
	}
	return it->value;
}

void delete(list *ls, int index) {
	if (index > ls->length-1) {
		// TODO: Error
		return;
	}

	item *it = ls->head;
	for (int i = 0; i < index; i++) {
		it = it->next;
	}


	if (it == ls->tail) {
		ls->tail = it->prev;
	} else {
		it->next->prev = it->prev;
		it->prev->next = it->next;
	}

	ls->length--;

	free(it);
}

void print(list *ls) {
	item *it = ls->head;

	for (int i = 0; i < ls->length; i++) {
		printf("%d ", it->value);
		it = it->next;
	}
	puts("");
}

int main() {
	list *ls = new_list(5);
	set(ls, 0, 10);
	set(ls, 4, 55);
	expand(ls, 2);
	set(ls, 2, 19);
	set(ls, 6, 66);
	delete(ls, 5);
	print(ls);
	printf("%d\n", get(ls, 2));

	return 0;
}

File: list.zig
const std = @import("std");

const Item = struct {
    val: i32,
    prev: ?*Item = null,
    next: ?*Item = null,

    fn new(allocator: std.mem.Allocator, val: i32) !*Item {
        var it = try allocator.create(Item);
        it.val = val;
        it.prev = null;
        it.next = null;
        return it;
    }

    fn print(self: *Item) void {
        std.debug.print("{d}\n", .{self.val});
    }
};

const ListErrors = error{
    OutOfBounds,
};

const List = struct {
    len: usize = 0,
    head: ?*Item = null,
    tail: ?*Item = null,
    allocator: std.mem.Allocator,

    fn new(allocator: std.mem.Allocator) !*List {
        var ls = try allocator.create(List);
        ls.allocator = allocator;
        ls.len = 0;
        ls.head = null;
        ls.tail = null;
        return ls;
    }

    fn append(self: *List, val: i32) !void {
        var it = try Item.new(self.allocator, val);

        if (self.len == 0) {
            self.head = it;
            self.tail = it;
        } else {
            self.tail.?.next = it;
            it.prev = self.tail;
            self.tail = it;
        }

        self.len += 1;
    }

    // fn insert(self: *List, index: usize, val: i32) !void {
    //     var it = try Item.new(self.allocator, val);

    //     self.len += 1;
    // }

    fn get(self: *List, index: usize) !*Item {
        if (index >= self.len) {
            return ListErrors.OutOfBounds;
        }

        var it = self.head;
        for (0..index) |_| {
            it = it.?.next;
        }
        return it.?;
    }

    fn set(self: *List, index: usize, val: i32) !void {
        if (index >= self.len) {
            return ListErrors.OutOfBounds;
        }

        var it = self.head;
        for (0..index) |_| {
            it = it.?.next;
        }
        it.?.val = val;
    }

    fn delete(self: *List, index: usize) !i32 {
        if (index >= self.len) {
            return ListErrors.OutOfBounds;
        }

        var it = self.head.?;
        for (0..index) |_| {
            it = it.next.?;
        }

        if (self.head == it) {
            self.head = it.next;
            self.head.?.prev = null;
        } else if (self.tail == it) {
            self.tail = it.prev;
            self.tail.?.next = null;
        } else {
            it.prev.?.next = it.next;
            it.next.?.prev = it.prev;
        }

        return it.val;
    }

    fn debug(self: *List) void {
        std.debug.print("{}\n", .{self});
    }

    fn print(self: *List) void {
        if (self.len == 0) {
            return;
        }

        var it = self.head;
        while (it != null) {
            std.debug.print("{d} ", .{it.?.val});
            it = it.?.next;
        }
        std.debug.print("\n", .{});
    }
};

pub fn main() !void {
    var arena = std.heap.ArenaAllocator.init(std.heap.page_allocator);
    defer arena.deinit();
    const allocator = arena.allocator();

    var ls = try List.new(allocator);
    try ls.append(5);
    try ls.append(3);
    ls.print();

    try ls.set(1, 91);
    try ls.append(0);
    try ls.append(7);
    const it = try ls.get(1);
    it.print();
    _ = try ls.delete(0);
    ls.print();
}
